shared/actions/provision.js

```diff
@@ -0,0 +1,438 @@
+// @flow
+import * as Constants from '../constants/provision'
+import * as RouteConstants from '../constants/route-tree'
+import * as DevicesGen from './devices-gen'
+import * as ProvisionGen from './provision-gen'
+import * as RPCTypes from '../constants/types/rpc-gen'
+import * as Saga from '../util/saga'
+import * as RouteTree from './route-tree'
+import * as Tabs from '../constants/tabs'
+import {isMobile} from '../constants/platform'
+import HiddenString from '../util/hidden-string'
+import {type TypedState} from '../constants/reducer'
+import {niceError} from '../util/errors'
+import {devicesTab as settingsDevicesTab} from '../constants/settings'
+import type {CommonResponseHandler} from '../engine/types'
+
+const devicesRoot = isMobile ? [Tabs.settingsTab, settingsDevicesTab] : [Tabs.devicesTab]
+
+type ValidCallbacks =
+  | 'keybase.1.gpgUi.selectKey'
+  | 'keybase.1.loginUi.displayPrimaryPaperKey'
+  | 'keybase.1.loginUi.getEmailOrUsername'
+  | 'keybase.1.provisionUi.DisplayAndPromptSecret'
+  | 'keybase.1.provisionUi.DisplaySecretExchanged'
+  | 'keybase.1.provisionUi.PromptNewDeviceName'
+  | 'keybase.1.provisionUi.ProvisioneeSuccess'
+  | 'keybase.1.provisionUi.ProvisionerSuccess'
+  | 'keybase.1.provisionUi.chooseDevice'
+  | 'keybase.1.provisionUi.chooseGPGMethod'
+  | 'keybase.1.secretUi.getPassphrase'
+
+const cancelOnCallback = (_: any, response: CommonResponseHandler, __: any) => {
+  response.error({code: RPCTypes.constantsStatusCode.scgeneric, desc: Constants.cancelDesc})
+}
+const ignoreCallback = (_: any, __: any) => {}
+
+// The provisioning flow is very stateful so we use a class to handle bookkeeping
+// We only allow one manager to be alive at a time
```

Comment:
Is anything guaranteeing this? You could make this an actual singleton and have a `reInit` method on it.

Reply:
i spent some time doing this but it makes the flow typing really crappy and complicated. This class isn't exported at all so its just used internally. I'm not very worried about someone making more of these things

Reply:
i'm making it a little harder to make an instance of this but not making it impossible